map/mode.c

```diff
@@ -7,69 +7,353 @@
+
+  SET_ITER(api->channels, entry, tmp) {
+    if (memcmp(trits_begin(mam_channel_id(&entry->value)), channel_id_trits, MAM_CHANNEL_ID_TRIT_SIZE) == 0) {
+      return &entry->value;
+    }
+  }
+
+  return NULL;
+}
+
+/**
+ * Gets an endpoint from its id
+ *
+ * @param api - The API [in]
+ * @param channel_id - The associated channel id [in]
+ * @param endpoint_id - The endpoint id [in]
+ *
+ * @return a pointer to the endpoint or NULL if not found
+ */
+static mam_endpoint_t *mam_api_endpoint_get(mam_api_t const *const api, tryte_t const *const channel_id,
+                                            tryte_t const *const endpoint_id) {
+  trit_t endpoint_id_trits[MAM_ENDPOINT_ID_TRIT_SIZE];
+  mam_endpoint_t_set_entry_t *entry = NULL;
+  mam_endpoint_t_set_entry_t *tmp = NULL;
+  mam_channel_t *parent_channel = mam_api_channel_get(api, channel_id);
+
+  if (endpoint_id == NULL || parent_channel == NULL) {
+    return NULL;
+  }
+
+  trytes_to_trits(endpoint_id, endpoint_id_trits, MAM_ENDPOINT_ID_TRYTE_SIZE);
+
+  SET_ITER(parent_channel->endpoints, entry, tmp) {
+    if (memcmp(trits_begin(mam_endpoint_id(&entry->value)), endpoint_id_trits, MAM_ENDPOINT_ID_TRIT_SIZE) == 0) {
+      return &entry->value;
+    }
+  }
+
+  return NULL;
+}
+
+/***********************************************************************************************************
+ * External functions
+ ***********************************************************************************************************/
 
-  mam_api_bundle_write_header_on_endpoint(api, channel_id, endpoint_id, NULL, NULL, bundle, msg_id);
+status_t map_mam_init(mam_api_t *const api, const iota_config_t *const iconf, tryte_t const *const seed,
+                      int32_t channel_ord, mam_psk_t_set_t *const psks, mam_ntru_pk_t_set_t *const ntru_pks,
+                      tryte_t const *const psk, tryte_t const *const ntru_pk) {
+  status_t ret = SC_OK;
+  if (!api || (!iconf && !seed)) {
+    return SC_MAM_NULL;
+  }
+
+  if (seed[0]) {
+    if (mam_api_init(api, seed) != RC_OK) {
+      ta_log_error("%s\n", "SC_MAM_FAILED_INIT");
+      return SC_MAM_FAILED_INIT;
+    }
+    api->channel_ord = channel_ord;
+  } else {
+    // Use local MAM file on the current machine
+    retcode_t rc = mam_api_load(iconf->mam_file_path, api, NULL, 0);
+    if (rc == RC_UTILS_FAILED_TO_OPEN_FILE) {
+      if (mam_api_init(api, (tryte_t *)iconf->seed) != RC_OK) {
+        ta_log_error("%s\n", "SC_MAM_FAILED_INIT");
+        return SC_MAM_FAILED_INIT;
+      }
+    } else if (rc != RC_OK) {
+      ta_log_error("%s\n", "SC_MAM_FAILED_INIT");
+      return SC_MAM_FAILED_INIT;
+    }
+  }
+
+  ret = key_set_setup(psks, ntru_pks, psk, ntru_pk);
+  if (ret) {
+    ta_log_error("%d\n", ret);
+    return ret;
+  }
+
+  return SC_OK;
+}
+
+status_t map_write_message_to_bundle_on_available_sk(const iota_client_service_t *const service, mam_api_t *const api,
```

Comment:
I am curious about the reason why the long function name `map_write_message_to_bundle_on_available_sk`.